### PR TITLE
Social distancing

### LINKED ISSUE #4: Social Distancing Detector Application
An application that detects pedestrians and identifies if two or more people are in close proximity and violate the physical distancing orders. 

The application can be based on current pedestrian detection models that run on edge TPUs and GPUs. 
E.g. [thatbrguy/Pedestrian-Detection](https://github.com/thatbrguy/Pedestrian-Detection)
pedestrian-detection

### PR DESCRIPTION
**Pull request by JsonSadler**
Hi @mhejrati , please review our first implementation.
Look into the proper Dockerfile, build it and run it. 

Any feedback and ideas are welcome.

### CODE REVIEW

File: social-distancing/Dockerfile-amd64-usbtpu [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+RUN apt-get update && apt-get install -y wget gnupg usbutils
+
+RUN echo "deb https://packages.cloud.google.com/apt coral-edgetpu-stable main" | tee /etc/apt/sources.list.d/coral-edgetpu.list
+RUN wget -qO - https://packages.cloud.google.com/apt/doc/apt-key.gpg | apt-key add -
+
+RUN apt-get update && apt-get install -y python3-pip pkg-config libedgetpu1-std python3-wget
+
+RUN python3 -m pip install https://dl.google.com/coral/python/tflite_runtime-2.1.0.post1-cp37-cp37m-linux_x86_64.whl 
+
+RUN apt-get install -y python3-flask
```
Comment (mhejrati) [inline social-distancing/Dockerfile-amd64-usbtpu:19]: I suggest merging these into one line.

File: social-distancing/libs/ConfigEngine.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
+#!/usr/bin/python3
```
Comment (mhejrati) [inline social-distancing/libs/ConfigEngine.py:1]: This looks nice. Thanks for providing the config engine. Can you please add documentation and example use cases for this?

File: social-distancing/neuralet-distancing.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
+#!/usr/bin/python3
+import sys
+
+import argparse
+
+from libs.Core import Distancing as CvEngine
+from libs.ConfigEngine import ConfigEngine
+from ui.WebGUI import WebGUI as UI
+
+class DistanceApp():
```
Comment (mhejrati) [inline social-distancing/neuralet-distancing.py:10]: Can you please add documentation for the DistanceApp class?

File: social-distancing/libs/Core.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
+import os, re
+import time
+
+import numpy as np
+import cv2 as cv
+from scipy.spatial import distance as dist
+from libs.centroid_object_tracker import CentroidTracker
+
+class Distancing:
```
Comment (mhejrati) [inline social-distancing/libs/Core.py:9]: Can you please document this class?

File: social-distancing/libs/Core.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+        return object_list list of  dict for each obj, 
+        obj["bbox"] is normalized coordinations for [x0, y0, x1, y1] of box
+        """
+        if self.device == 'Dummy': 
+            return cv_image, [], None
+
+        resized_image = cv.resize(cv_image, tuple(self.image_size[:2]))
+        rgb_resized_image = cv.cvtColor(resized_image, cv.COLOR_BGR2RGB)
+        tmp_objects_list = self.detector.inference(rgb_resized_image)
+        hscale = cv_image.shape[0]/resized_image.shape[0]
```
Comment (mhejrati) [inline social-distancing/libs/Core.py:48]: can you please remove hscale and wscale variables? They don't seem to be used anywhere in this function.

File: social-distancing/libs/Core.py [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+            y0 = box[0]
+            x1 = box[3]
+            y1 = box[2]
+            obj["centroid"] = [(x0+x1)/2, (y0+y1)/2, x1 - x0, y1 - y0]
+            obj["bbox"] = [x0, y0, x1, y1]
+
+        objects_list, distancings = self.calculate_distancing(tmp_objects_list)
+        return cv_image, objects_list, distancings
+
+    def process_video(self, video_uri):
```
Comment (mhejrati) [inline social-distancing/libs/Core.py:63]: Can you please document this function? What are acceptable URI inputs?

File: social-distancing/libs/Core.py [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+        self.running_video = True
```
Comment (mhejrati) [inline social-distancing/libs/Core.py:64]: Why is running_video set to True here? Seems to need to move before the while loop.

File: social-distancing/libs/Core.py [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+            print('opened video ', video_uri)
+        else:
+            print('failed to load video ', video_uri)
+            return 
+
+        while input_cap.isOpened() and self.running_video:
+            _, cv_image = input_cap.read()
+            _, objects, distancings = self.__process(cv_image)
+            self.ui.update(cv_image, objects, distancings)
+            time.sleep(0.030)
```
Comment (mhejrati) [inline social-distancing/libs/Core.py:77]: Why do we need this sleep?

File: social-distancing/libs/centroid_object_tracker.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
+# import the necessary packages
+from scipy.spatial import distance as dist
+from collections import OrderedDict
+import numpy as np
+
+class CentroidTracker():
```
Comment (mhejrati) [inline social-distancing/libs/centroid_object_tracker.py:6]: I suggest using a consistent naming convention. We like to use python black to lint the codebase.

File: social-distancing/libs/detectors/edgetpu/MobileNetSSD.py [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+import time
+from threading import Thread
+from queue import Queue
+import numpy as np
+import wget 
+
+from tflite_runtime.interpreter import load_delegate
+from tflite_runtime.interpreter import Interpreter
+
+class Detector():
```
Comment (mhejrati) [inline social-distancing/libs/detectors/edgetpu/MobileNetSSD.py:12]: Can you please document this class? It looks very similar to Neuralet source code.

File: social-distancing/libs/detectors/edgetpu/MobileNetSSD.py [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+        self.interpreter.invoke()
+        inference_time = time.perf_counter() - t_begin
+        boxes = self.interpreter.get_tensor(self.output_details[0]['index'])
+        labels = self.interpreter.get_tensor(self.output_details[1]['index'])
+        scores = self.interpreter.get_tensor(self.output_details[2]['index'])
+        num = self.interpreter.get_tensor(self.output_details[3]['index'])
+        
+        result = []
+        for i in range(boxes.shape[1]): #number of boxes
+            if labels[0, i] == self.class_id and scores[0, i] > self.score_threshold:
```
Comment (mhejrati) [inline social-distancing/libs/detectors/edgetpu/MobileNetSSD.py:54]: Where is the score_threshold defined?

File: social-distancing/ui/WebGUI.py [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+
+            # yield the output frame in the byte format
+            yield (
+                b"--frame\r\n"
+                b"Content-Type: image/jpeg\r\n\r\n" + bytearray(encodedImage) + b"\r\n"
+            )
+
+    def _run(self):
+        self.app.run(
+            host="0.0.0.0", port=8000, debug=True, threaded=True, use_reloader=False,
```
Comment (mhejrati) [inline social-distancing/ui/WebGUI.py:101]: It would be great to read this from the config engine.

File: social-distancing/ui/WebGUI.py [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+            use_normalized_coordinates=True,
+            line_thickness=3,
+        )
+        with self._lock:
+            self._output_frame = input_frame.copy()
+
+    def create_flask_app(self):
+        app = Flask(__name__)
+
+        # @staticmethod
```
Comment (mhejrati) [inline social-distancing/ui/WebGUI.py:64]: Please remove.

**Review by mhejrati — APPROVED** [review]
Many thanks for such great PR. This looks awesome. 

There are some minor comments that I added, I will merge but hope we can fix them soon.

File: social-distancing/ui/WebGUI.py (hunk 12 above)
Comment (mrn-mln) [reply social-distancing/ui/WebGUI.py:101]: it's fixed, I will send a PR for you.

**Comment by JsonSadler** [comment]
WOW! 
I was away and had no idea that you have merged, and the project has been so active! 
I made this PR for fun and I am not sure if I have enough time to review the code as it seems to had lots of changes. I am happy my contribution was helpful. hope this helps countries to handle the coronavirus better.

Thanks for your great project.

**Comment by rkgits** [comment]
Thank you @JsonSadler! 
This is an open-source project that can save lives. We are actively working on this project for the past 8 weeks and we really appreciate any kind of contribution. 
We definitely need some UI help at this moment, I appreciate it if you can contribute or promote this project in developer communities so we can get help from other awesome individuals like you!

**Comment by rkgits** [comment]
P.S.: We will create issues for UI works/stories soon. I will share the link with you, feel free to share it with others.

**Pull request merged**